docs/design/2022-09-28-flashback-to-timestamp.md

```diff
@@ -0,0 +1,120 @@
+// make sure the region is "locked" by `PrepareFlashbackToVersionRequest` first,
+// otherwise this request will fail.
+message FlashbackToVersionRequest {
+    Context context = 1;
+    // The TS version which the data should flashback to.
+    uint64 version = 2;
+    bytes start_key = 3;
+    bytes end_key = 4;
+    // The `start_ts`` and `commit_ts` which the newly written MVCC version will use.
+    uint64 start_ts = 5;
+    uint64 commit_ts = 6;
+}
+
+message FlashbackToVersionResponse {
+    errorpb.Error region_error = 1;
+    string error = 2;
+}
+```
+
+Then a `Flashback To Timestamp` DDL job can be simply divided into the following steps
+
+* Save values of some global variables and PD schedule. Those values will be changed during `Flashback`.
+
+* Pre-checks. After all checks are passed, TiDB will disable GC and closed PD schedule for the cluster. The specific checks are as follows:
+    * The FlashbackTS is after `tikv_gc_safe_point`.
+    * The FlashbackTS is before the minimal store resolved TS.
+    * No related DDL history in flashback time range.
+    * No running related DDL jobs.
+
+* TiDB get flashback key ranges and splits them into separate regions to avoid locking unrelated key ranges. Then TiDB send `PrepareFlashbackToVersion` RPC requests to lock regions in TiKV. Once locked, no more read, write and scheduling operations are allowed for those regions.
+
+* After locked all relevant key ranges, the DDL Owner will update schema version and synchronize it to other TiDBs. When other TiDB applies the `SchemaDiff` of type `Flashback To Timestamp`, it will disconnect all relevant links.
+
+* Send `FlashbackToVersion` RPC requests to all relevant key ranges with same `commit_ts`. Each region handles its own flashback progress independently.
+    * Read the old MVCC data and write it again with the given `commit_ts` to pretend it's a new transaction commit.
+    * Release the Raft proposing lock and resume the lease read.
+
+* TiDB checks whether all the requests returned successfully, and retries those that failed with same `commit_ts` until the whole flashback is done.
+
+* After `Flashback To Timestamp` is finished, TiDB will restore all changed global variables and restart PD schedule. At the same time, notify `Stats Handle` to reload statistics from TiKV.
+
+### New Syntax Overview
+
+TiDB will support 3 new syntaxes as follows.
+
+1. Flashback whole cluster except some system tables to the specified timestamp.
+
+```sql
+FLASHBACK CLUSTER TO TIMESTAMP '2022-07-05 08:00:00';
+```
+
+2. Flashback some databases to the specified timestamp.
+
+```sql
+FLASHBACK DATABASE [db1], [db2] TO TIMESTAMP '2022-07-05 08:00:00';
+```
+
+3. Flashback some tables to the specified timestamp.
+
+```sql
+FLASHBACK TABLE [table1], [table2] TO TIMESTAMP '2022-08-10 08:00:00';
+```
+
+### Limitations and future Work
+
+1. DDL history exists for the flashback time period is not currently supported, the error message is shown below:
```

Comment:
This limitation make sense from the implementation side consideration.
Although for the cluster scope, no DDL during that period restriction is still a bit strong.

Reply:
Yes, try to improve it in the future.